src/components/backend-ai-environment-list.ts

```diff
@@ -431,41 +432,81 @@ export default class BackendAIEnvironmentList extends BackendAIPage {
     }
   }
 
+  /**
+   * Validate backend.ai service ports.
+   */
+  _isServicePortValid() {
+    const container = this.shadowRoot.querySelector("#modify-app-container");
+    const rows = container.querySelectorAll(".row:not(.header)");
+    const ports = new Set();
+    for (const row of rows) {
+      const textFields = row.querySelectorAll("wl-textfield");
+      if (Array.prototype.every.call(textFields, field => field.value === "")) {
+        continue;
+      }
+
+      const appName = textFields[0].value, protocol = textFields[1].value, port = parseInt(textFields[2].value);
+      if (appName === "") {
+        this.servicePortsMsg = '*App Name may not be empty.'
+        return false;
+      }
+      if (!["http", "tcp", "pty"].includes(protocol)) {
+        this.servicePortsMsg = '*Protocol must be one of "http", "tcp", "pty".';
```

Comment:
The message should use i18n methods. e.g. `_t('environment.ProtocolMustBeOneOfHTTPTCPPTY')`

Reply:
Also, special characters such as asterisk should be avoided.